MNT: bump python versions and dependency minimum versions

rcomer opened the pull request:
<!--

Thanks for contributing to cartopy!
Please use this template as a guide to streamline the pull request you are about to make.

Remember: it is significantly easier to merge small pull requests. Consider whether this pull
request could be broken into smaller parts before submitting.

-->



## Rationale

<!-- Please provide detail as to *why* you are making this change. -->
Python 3.8 recently fell off the NEP29 schedule, so this PR removes testing with that and also adds it for python 3.11.
https://numpy.org/neps/nep-0029-deprecation_policy.html

I also bumped versions of other packages used in the minimum versions test as I assume bad things can happen if you bump the python version without doing this!  I wasn't sure exactly which versions to use here but mostly went for the oldest that is newer than 2 years.  For shapely it seemed safer to go with the older 1.7.1 (August 2020) as 1.8.x had many patch releases stretching into last year.  Matplotlib 3.4 is also older than 2 years but 3.5 didn't come out till November 2021.

## Implications

<!-- If applicable, to the best of your knowledge, what are the implications of this change? -->


<!--
## Checklist

 * If you have not already done so, ensure you've read and signed the Contributor Licence Agreement (CLA).
   (See the [governance page](http://scitools.org.uk/governance.html) for the CLA and what to do with it).

 * If this is a new feature, please provide an example of its use in the description. We may want to make a
   follow-on pull request to put the example in the gallery!

 * Ensure there is a suitable item in the cartopy test suite for the change you are proposing.

-->

greglucas reviewed (approved):
It looks like there is one `MPL_VERSION < (3, 3)` check in the tests that could be dropped too. Are we also supposed to update the build-system dependencies in `pyproject.toml`?

rcomer commented:
> Are we also supposed to update the build-system dependencies in pyproject.toml?

That seems likely (I'm by no means an expert on this!)  I updated the cython version here to match the min version tests.  The setuptools version here is from 2018 and setuptools_scm only last year, so not sure they were previously following the ~2 year policy.

I removed the mpl version check from that test, and also found some handling for older mpl versions in the main code.

greglucas commented on lib/cartopy/mpl/geoaxes.py:
```diff
@@ -1783,9 +1783,7 @@ def _wrap_args(self, *args, **kwargs):
-        # The shading keyword argument was added in MPL 3.3, so keep
-        # this default updating until we only support MPL>=3.3
-        default_shading = mpl.rcParams.get('pcolor.shading', 'auto')
+        default_shading = mpl.rcParams.get('pcolor.shading')
```
I'm not sure how to read the above comment... Can we get rid of the default altogether now and assume kwargs contains "shading" now?

```suggestion
```

rcomer replied:
I think the user can just not pass _shading_, or am I misunderstanding the question?

rcomer commented on lib/cartopy/mpl/geoaxes.py:
```diff
@@ -44,7 +45,7 @@
 
 
-assert mpl.__version__ >= '3.4', \
+assert packaging.version.parse(mpl.__version__).release[:2] >= (3, 4), \
```
I guess mpl v3.10 will ship next year, at which point the string comparison would become problematic.
